iree/compiler/Dialect/HAL/Target/VMLA/VMLATarget.cpp

```diff
@@ -66,6 +90,191 @@ class VMLATargetBackend final : public TargetBackend {
+          // SymbolTable::replaceAllSymbolUses only looks at root symbols,
+          // which we can't blindly replace (other targets will map to other
+          // linked executables).
+          auto executableUses =
+              SymbolTable::getSymbolUses(executableOp, moduleOp);
+          if (!executableUses.hasValue()) continue;
+          for (auto executableUse : executableUses.getValue()) {
+            auto executableUser = executableUse.getUser();
+            // Only process symbols for this @target::@entry.
+            auto nestedRefs =
+                executableUse.getSymbolRef().getNestedReferences();
+            if (nestedRefs.size() != 2 ||
+                nestedRefs[0].getValue() != targetOp.sym_name() ||
+                nestedRefs[1].getValue() != entryPointOp.sym_name()) {
+              continue;
+            }
+            if (auto dispatchOp =
+                    dyn_cast<IREE::HAL::CommandBufferDispatchSymbolOp>(
+                        executableUser)) {
+              // New nested reference to the linked exe/target/entry.
+              StringRef newExecutableOpSymName =
+                  linkedExecutableOp
+                      .getAttrOfType<StringAttr>(
+                          SymbolTable::getSymbolAttrName())
+                      .getValue();
+              auto newSymbolRefAttr = builder.getSymbolRefAttr(
+                  newExecutableOpSymName,
+                  {builder.getSymbolRefAttr(linkedTargetOp),
+                   builder.getSymbolRefAttr(newEntryPointOp)});
+              dispatchOp.setAttr("entry_point", newSymbolRefAttr);
+            }
+          }
+        }
+
+        // Clone vm.module ops, including their contents.
+        // As we do this, we de-dup some symbols.
+        auto vmModuleOps =
+            targetOp.getInnerModule().getOps<IREE::VM::ModuleOp>();
+        if (vmModuleOps.empty()) {
+          return targetOp.getInnerModule().emitError()
+                 << "target's outer module does not contain a vm.module op";
+        }
+        auto vmModuleOp = *vmModuleOps.begin();
+        builder.setInsertionPoint(&linkedVmModuleOp.getBlock().back());
+        SymbolTable symbolTable(linkedVmModuleOp.getOperation());
+
+        for (auto funcOp : vmModuleOp.getOps<IREE::VM::FuncOp>()) {
+          builder.clone(*funcOp);
+        }
+        for (auto exportOp : vmModuleOp.getOps<IREE::VM::ExportOp>()) {
+          builder.clone(*exportOp);
+        }
+
+#define IREE_CLONE_OP_WITHOUT_DUPLICATES(opType)          \
+  for (auto op : vmModuleOp.getOps<IREE::VM::opType>()) { \
+    if (symbolTable.lookup(op.getName())) continue;       \
+    builder.clone(*op);                                   \
+  }
+        IREE_CLONE_OP_WITHOUT_DUPLICATES(ImportOp);
+        IREE_CLONE_OP_WITHOUT_DUPLICATES(RodataOp);
+        IREE_CLONE_OP_WITHOUT_DUPLICATES(GlobalI32Op);
+        IREE_CLONE_OP_WITHOUT_DUPLICATES(GlobalI64Op);
+        IREE_CLONE_OP_WITHOUT_DUPLICATES(GlobalRefOp);
```

Comment:
This block of code is sketchy to me, wondering if there's a better way to merge the contents of the current executable into the linked executable:

* With some concept of operation ordering/sorting
* Handling duplicate symbols intelligently (imports are unique, but what about globals? should those get new names instead of being skipped?)
* Using generics instead of listing ops explicitly?

Reply:
you could just do this for all top level symbols - you can use hasTrait<Symbol> on each op and then do the dedupe. what's allowed in a vm.module is a constrained set and it's nice to use traits so that when new ops are added someone won't have to fix this.

Reply:
👍 done, `hasTrait` works nicely here